### PR TITLE
Kernel: Add an initial implementation of virtio-net driver.

### PR DESCRIPTION
**Pull request by nichtverstehen**
It can be exercised by setting `SERENITY_ETHERNET_DEVICE_TYPE=virtio-net-pci`.

### CODE REVIEW

File: Kernel/Net/VirtIO/VirtIONetworkAdapter.cpp [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+static constexpr u8 VIRTIO_NET_HDR_F_DATA_VALID = 1;
+static constexpr u8 VIRTIO_NET_HDR_F_RSC_INFO = 1;
+static constexpr u8 VIRTIO_NET_HDR_GSO_NONE = 0;
+static constexpr u8 VIRTIO_NET_HDR_GSO_TCPV4 = 1;
+static constexpr u8 VIRTIO_NET_HDR_GSO_UDP = 3;
+static constexpr u8 VIRTIO_NET_HDR_GSO_TCPV6 = 4;
+static constexpr u8 VIRTIO_NET_HDR_GSO_UDP_L4 = 5;
+static constexpr u8 VIRTIO_NET_HDR_GSO_ECN = 0x80;
+
+struct [[gnu::packed]] virtio_net_config {
```
Comment (gmta) [inline Kernel/Net/VirtIO/VirtIONetworkAdapter.cpp:58]: These two structs lend their names directly from the spec, but our coding style dictates a camel case name for structs and classes. I think that as long as the name is largely the same, it's still clear what part of the spec it refers to.

File: Kernel/Net/VirtIO/VirtIONetworkAdapter.cpp [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+ * Copyright (c) 2023, Kirill Nikolaev <cyril7@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Bus/PCI/IDs.h>
+#include <Kernel/Net/NetworkingManagement.h>
+#include <Kernel/Net/VirtIO/VirtIONetworkAdapter.h>
+
+namespace virtio {
```
Comment (gmta) [inline Kernel/Net/VirtIO/VirtIONetworkAdapter.cpp:12]: Our namespaces generally start with an uppercase letter; i.e. `namespace VirtIO`

File: Kernel/Net/VirtIO/VirtIONetworkAdapter.cpp [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+            negotiated |= VIRTIO_NET_F_MTU;
+        return negotiated;
+    });
+    if (!success)
+        return Error::from_errno(EIO);
+
+    success = handle_device_config_change();
+    if (!success) {
+        return Error::from_errno(EIO);
+    }
```
Comment (gmta) [inline Kernel/Net/VirtIO/VirtIONetworkAdapter.cpp:147]: You can drop the braces for single-line if-blocks like these.

File: Kernel/Net/VirtIO/VirtIONetworkAdapter.cpp [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+{
+    UserOrKernelBuffer buf = UserOrKernelBuffer::for_kernel_buffer(const_cast<u8*>(data));
+
+    size_t offset = 0;
+    while (offset < length) {
+        PhysicalAddress start_of_chunk;
+        size_t length_of_chunk;
+        VERIFY(ring.copy_data_in(buf, offset, length - offset, start_of_chunk, length_of_chunk));
+        if (!chain.add_buffer_to_chain(start_of_chunk, length_of_chunk, VirtIO::BufferType::DeviceReadable)) {
+            // TODO: Should rewind RingBuffer?
```
Comment (gmta) [inline Kernel/Net/VirtIO/VirtIONetworkAdapter.cpp:253]: Please formulate a concise `// FIXME:` comment instead as per our coding style.

File: Kernel/Net/VirtIO/VirtIONetworkAdapter.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2021, the SerenityOS developers.
```
Comment (gmta) [inline Kernel/Net/VirtIO/VirtIONetworkAdapter.h:2]: Why add this 2021 copyright? Your own copyright is enough :^)

File: Kernel/Net/VirtIO/VirtIONetworkAdapter.h [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    // VirtIO::Device
+    virtual bool handle_device_config_change() override;
+    virtual void handle_queue_update(u16 queue_index) override;
+
+    // NetworkAdapter
+    virtual void send_raw(ReadonlyBytes) override;
+
+private:
+    VirtIO::Configuration const* m_device_config { nullptr };
+    // TODO: Need to be atomic as they are read without sync?
```
Comment (gmta) [inline Kernel/Net/VirtIO/VirtIONetworkAdapter.h:50]: Please formulate a `// FIXME:`

File: Kernel/Net/VirtIO/VirtIONetworkAdapter.cpp [hunk 7]
```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (c) 2023, the SerenityOS developers.
```
Comment (gmta) [inline Kernel/Net/VirtIO/VirtIONetworkAdapter.cpp:2]: Your own copyright is enough.

**Comment by nichtverstehen** [comment]
Thanks for the comments, Jelle! Amended and pushed.

File: Kernel/Net/VirtIO/VirtIONetworkAdapter.cpp [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+    if (!success)
+        return Error::from_errno(EIO);
+
+    success = handle_device_config_change();
+    if (!success)
+        return Error::from_errno(EIO);
+
+    success = setup_queues(2); // receive & transmit
+    if (!success)
+        return Error::from_errno(EIO);
```
Comment (ADKaster) [inline Kernel/Net/VirtIO/VirtIONetworkAdapter.cpp:149]: This pattern is quite meh. A follow-up change to update the base class's virtual methods here to return `ErrorOr<void>` would be a nice Kernel refactoring PR. That way we could just TRY() these methods.

**Review by ADKaster — APPROVED** [review]
This looks like a pretty sweet driver change! A follow-up PR to update `Meta/run.sh`  to enable this adaptor alongside the other VirtIO devices would be much appreciated as well. Not sure if they're hidden behind an environment variable flag or qemu feature detection or what, though.

**Comment by nichtverstehen** [comment]
Yeah, enabling this adapter is pretty easy:
```
SERENITY_ETHERNET_DEVICE_TYPE=virtio-net-pci Meta/serenity.sh run
```
Granted, it's not very discoverable. However most users would not care to change the network implementation?..

@ADKaster do you know of a good place in the docs where I could mention this option?

**Pull request merged**